client/my-sites/domains/domain-management/transfer/transfer-to-other-site/confirmation-dialog.jsx

```diff
@@ -11,25 +11,80 @@ import { get } from 'lodash';
 /**
  * Internal Dependencies
  */
-import { getSite } from 'calypso/state/sites/selectors';
+import { currentUserHasFlag } from 'calypso/state/current-user/selectors';
+import { getSite, isCurrentPlanPaid } from 'calypso/state/sites/selectors';
 import { Dialog } from '@automattic/components';
+import { NON_PRIMARY_DOMAINS_TO_FREE_USERS } from 'calypso/state/current-user/constants';
 
 class TransferConfirmationDialog extends React.PureComponent {
 	static propTypes = {
+		isMapping: PropTypes.bool.isRequired,
 		isVisible: PropTypes.bool.isRequired,
 		targetSiteId: PropTypes.number.isRequired,
 		disableDialogButtons: PropTypes.bool.isRequired,
 		domainName: PropTypes.string.isRequired,
 		onConfirmTransfer: PropTypes.func.isRequired,
 		onClose: PropTypes.func.isRequired,
+		primaryWithPlansOnly: PropTypes.bool,
 	};
 
 	onConfirm = ( closeDialog ) => {
 		this.props.onConfirmTransfer( this.props.targetSite, closeDialog );
 	};
 
+	getMessage() {
+		const { domainName, isMapping, targetSite, translate } = this.props;
+		const targetSiteTitle = get( targetSite, 'title', translate( 'Site Title' ) );
+		if ( isMapping ) {
+			return translate(
+				'Do you want to transfer mapping of {{strong}}%(domainName)s{{/strong}} ' +
+					'to site {{strong}}%(targetSiteTitle)s{{/strong}}?',
+				{
+					args: { domainName, targetSiteTitle },
+					components: { strong: <strong /> },
+				}
+			);
+		}
+
+		return translate(
+			'Do you want to transfer {{strong}}%(domainName)s{{/strong}} ' +
+				'to site {{strong}}%(targetSiteTitle)s{{/strong}}?',
+			{
+				args: { domainName, targetSiteTitle },
+				components: { strong: <strong /> },
+			}
+		);
+	}
+
+	renderTargetSiteOnFreePlanWarnings() {
+		if ( this.props.primaryWithPlansOnly ) {
+			return (
+				<p>
+					{ this.props.translate(
+						"The target site doesn't have a paid plan, so you'll have to pay the full price for a " +
+							'domain mapping subscription when the domain mapping next renews. You will not be able to set it as primary either. ' +
+							'Consider upgrading the target site to a paid plan to get these features for free.'
+					) }
+				</p>
+			);
+		}
+
+		return (
+			<p>
+				{ this.props.translate(
+					"The target site doesn't have a paid plan, so you'll have to pay the full price for a " +
+						'domain mapping subscription when the domain mapping next renews. ' +
+						'Consider upgrading the target site to a paid plan to get domain mappings and many more features for free.'
```

Comment:
I don't like the wording here and above where we say the user gets "features for free", as they are definitely not free. 😬
How about something like the following?
> If you upgrade the target site to a paid plan, domain mappings (and many more features!) are included in our plans.

and

> If you upgrade the target site to a paid plan, these features are included in the plan.